x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -208,47 +219,118 @@ static void getMemoryAndProcessors(
+            int numberOfAvailableProcessors = numberOfAvailableProcessors(
+                autoscalingContext,
+                allocatedProcessorsScale,
+                processorsNewlyAssignedToTrainedModelsAllocationsThisTime
+            );
+            int numExistingProcessorsToBeUsed = Math.min(numProcessorsNeeded, numberOfAvailableProcessors);
 
             if (AssignmentState.STARTING.equals(assignment.getAssignmentState()) && assignment.getNodeRoutingTable().isEmpty()) {
-
+                // this state occurs when the assignment has just been added and has not been assigned to any nodes
                 logger.debug(
                     () -> format(
                         "trained model [%s] lacks assignment , memory required [%d]",
                         modelAssignment.getKey(),
                         estimatedMemoryUsage
                     )
                 );
-
-                extraSingleNodeModelMemoryInBytes = Math.max(extraSingleNodeModelMemoryInBytes, estimatedMemoryUsage);
+                extraPerNodeModelMemoryBytes = Math.max(extraPerNodeModelMemoryBytes, estimatedMemoryUsage);
                 extraModelMemoryInBytes += estimatedMemoryUsage;
 
-                // if not low priority, check processor requirements
+                // if not low priority, check processor requirements. If
                 if (Priority.LOW.equals(modelAssignment.getValue().getTaskParams().getPriority()) == false) {
-                    // as assignments can be placed on different nodes, we only need numberOfThreadsPerAllocation here
-                    extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, numberOfThreadsPerAllocation);
-                    extraProcessors += numberOfAllocations * numberOfThreadsPerAllocation;
+                    if (numProcessorsNeeded > numberOfAvailableProcessors) {
+                        // as assignments can be placed on different nodes, we only need numberOfThreadsPerAllocation here
+                        extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, numberOfThreadsPerAllocation);
+                        extraProcessors += numberOfRequestedAllocations * numberOfThreadsPerAllocation - numExistingProcessorsToBeUsed;
+                    }
+                    processorsNewlyAssignedToTrainedModelsAllocationsThisTime += numExistingProcessorsToBeUsed;
                 }
             } else if (assignment.getNodeRoutingTable().values().stream().allMatch(r -> r.getState().consumesMemory() == false)) {
-                // Ignore states that don't consume memory, for example all allocations are failed
+                // Ignore states that don't consume memory, for example all allocations are failed or stopped
                 continue;
+            } else if (AssignmentState.STARTED.equals(assignment.getAssignmentState()) && numMissingAllocations > 0) {
+                // this state occurs when we have a partial allocation, but not fully allocated
```

Comment:
Interesting, I'll explore this

Reply:
I think the difference between these two is that we don't update perNodeModelMemoryInBytes during the STARTING case, but I'm not sure thats the way it should be.

Reply:
I think that you are right though, it would be easier if we combined these two cases, and then we can add additional cases after this if-else if we need to.

Reply:
It would be good to merge the 2 cases, as far as I can see the main difference is in the particular message that is logged. 

This branch does not have the check for `Priority.LOW` deployments. Is that because a low priority deployment that is started cannot change the number of allocations it has